esrally/track/loader.py

```diff
@@ -468,7 +475,62 @@ def prepare_bundled_document_set(self, document_set, data_root):
                 return False
 
 
-def render_template(loader, template_name, template_vars=None, glob_helper=lambda f: [], clock=time.Clock):
+class TemplateSource:
+    """
+    Prepares the fully assembled track file from file or string.
+    Doesn't render using jinja2, but embeds track fragments referenced with
+    rally.collect(parts=...
+    """
+
+    collect_parts_re = re.compile(r'''{{\ +?rally\.collect\(parts="(.+?(?="))"\)\ +?}}''')
+
+    def __init__(self, base_path, template_file_name, source=io.FileSource, fileglobber=glob.glob):
+        self.base_path = base_path
+        self.template_file_name = template_file_name
+        self.source = source
+        self.fileglobber = fileglobber
+        self.assembled_source = None
+        self.logger = logging.getLogger(__name__)
+
+    def load_template_from_file(self):
+        loader = jinja2.FileSystemLoader(self.base_path)
+        try:
+            base_track = loader.get_source(jinja2.Environment(), self.template_file_name)
+        except jinja2.TemplateNotFound:
+            self.logger.exception("Could not track from [%s].", self.template_file_name)
+            raise TrackSyntaxError("Could not load track from '{}'".format(self.template_file_name))
+        self.assembled_source = self.replace_includes(self.base_path, base_track[0])
+
+    def load_template_from_string(self, template_source):
+        self.assembled_source = self.replace_includes(self.base_path, template_source)
+
+    def replace_includes(self, base_path, track_fragment):
+        match = TemplateSource.collect_parts_re.findall(track_fragment)
+        if match:
+            # Construct replacement dict for matched captures
+            repl = {}
+            for glob_pattern in match:
+                full_glob_path = os.path.join(base_path, glob_pattern)
+                sub_source = self.read_glob_files(full_glob_path)
+                repl[glob_pattern] = self.replace_includes(base_path=io.dirname(full_glob_path), track_fragment=sub_source)
+
+            def replstring(matchobj):
```

Comment:
I'd like to preserve the nested function to avoid complicating the parameters in `re.sub()` and keeping a useful comment about the matchobject which is hard to find in the Python docs.